src/main/scala/dsls/regex/RegularExpression.scala

```diff
@@ -11,6 +11,30 @@ package dsls.regex
 abstract class RegularExpression {
   /** returns true if the given string matches this regular expression */
   def matches(string: String) = RegexMatcher.matches(string, this)
+  def ||(right: RegularExpression): Union =
+    Union(this,right)
+  def ~(right: RegularExpression): Concat =
+    Concat(this,right)
+  def <*>(): Star =
+    Star(this)
+  def <+>(): Concat =
+    Concat(this, Star(this))
```

Comment:
Once you've defined the `~` operator, you can use it in your other functions (though you don't have to).  I wrote `<+>` as `this ~ (this <*>)`, but both implementations work, and I think yours might even be clearer.